### PR TITLE
Save inputs on change (not just blur)

### PR DESCRIPTION
**Pull request by olemartinorg**
## Description
Until now, changes in `InputComponent`, `TextAreaComponent` and `AddressComponent` were only saved server-side when the user defocuses/blurs the input field. This also meant that validation runs at the point when the user finishes filling out a certain field in the form. This change adds a hook to automatically save data (and validate) ~400ms after the last of (usually multiple) `change` events, effectively saving and validating after the user has stopped typing for a short while.

As the 'required'/empty field validation can get quite annoying if it pops up while you're focusing on a field and thinking of what to type, this specific validation has been postponed to run at the point when full page validation runs. Other validations will run as normal when the user stops typing.

## Related Issue(s)
- #98
- #155
- Altinn/altinn-cdn/pull/104
- Altinn/altinn-studio-docs/pull/580

## Reading material regarding debounce timeout:
- https://ux.stackexchange.com/questions/95336/how-long-should-the-debounce-timeout-be
- https://lawsofux.com/doherty-threshold/

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green
- [x] I have searched for other relevant issues regarding onBlur and closed/marked as duplicate/mentioned this change

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

**Comment by olemartinorg** [comment]
Presented a demo of this on friday, and the most important item of feedback received was that it should be possible to disable this altogether (or possibly adjust the debounce timeout) per-component. Will look into that before opening for review.

**Review by nkylstad — APPROVED** [review]
Looks great! 🦚 
I commented on a couple of small details :)

File: src/altinn-app-frontend/src/components/advanced/AddressComponent.tsx [hunk 1]
```diff
@@ -95,15 +123,15 @@ export function AddressComponent({
...
     (key: AddressKeys, value: any) => {
       const validationErrors: IAddressValidationErrors = validate();
       setValidations(validationErrors);
       if (!validationErrors[key]) {
-        handleDataChange(value, key);
+        handleDataChange(value, key, false, false);
         if (key === AddressKeys.zipCode && !value) {
           // if we are removing a zip code, also remove post place from form data
-          onBlurField(AddressKeys.postPlace, '');
+          setPostPlace('', true);
```
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/altinn-app-frontend/src/components/advanced/AddressComponent.tsx:134]: Good eye! It used to be that onBlur also stored the value, but that was changed up when it didn't work properly with formatting for `InputComponent`. I'll re-test and fix it.

File: src/altinn-app-frontend/src/components/hooks/useDelayedSavedState.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  const [immediateState, setImmediateState] = React.useState(formValue);
+  const [saveNextChangeImmediately, setSaveNextChangeImmediately] =
+    React.useState(false);
+
+  React.useEffect(() => {
+    setImmediateState(formValue);
+  }, [formValue]);
+
+  React.useEffect(() => {
+    if (typeof saveAfter === 'boolean' && !saveAfter) {
```
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/altinn-app-frontend/src/components/hooks/useDelayedSavedState.ts]: Smart! I tend to be very scared of implicit type coercion, so I always check very verbosely. You're right, this is much cleaner!

File: src/altinn-app-frontend/src/components/advanced/AddressComponent.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/altinn-app-frontend/src/components/advanced/AddressComponent.tsx:134]: D'oh! And my eye is not that great. missed which was the old line and which was the new one. Yes, setPostPlace() now stores the value both in local state, but also saves it to the server (after a timeout). With the second argument being true, it is set to being saved immediately.

**Comment by olemartinorg** [comment]
@nkylstad Thanks for the review! Could you have a look at the cdn and docs PRs as well? They should be merged when this is released, at least.

**Pull request merged**